include/seastar/parquet/file_reader.hh

```diff
@@ -0,0 +1,475 @@
+    // View the next page. Returns an empty result on eof.
+    seastar::future<std::optional<page>> next_page(uint32_t expected_header_size=_default_expected_header_size);
+};
+
+////////////////////////////////////////////////////////////////////////////////
+
+// A uniform interface to multiple compression algorithms.
+class decompressor {
+    const format::CompressionCodec::type _codec;
+    buffer _buffer;
+public:
+    explicit decompressor(format::CompressionCodec::type codec) : _codec{codec} {}
+    // View the decompressed data. Will throw if the size of decompressed data doesn't match decompressed_len.
+    std::basic_string_view<uint8_t> operator()(std::basic_string_view<uint8_t> compressed, size_t decompressed_len);
+};
+
+////////////////////////////////////////////////////////////////////////////////
+
+/* There are two encodings used for definition and repetition levels: RLE and BIT_PACKED.
+ * level_decoder provides a uniform interface to them.
+ */
+class level_decoder {
+    std::variant<RleDecoder, BitReader> _decoder;
+    int _bit_width;
+    int _num_values;
+    int _values_read;
+    int bit_width(int max_n) {
+        assert (max_n >= 0);
+        return (max_n == 0) ? 0 : (32 - __builtin_clz(static_cast<int32_t>(max_n)));
+    }
+public:
+    explicit level_decoder(int max_level) : _bit_width(bit_width(max_level)) {}
+
+    // Set a new source of levels. V1 and V2 are for data pages V1 and V2 respectively.
+    // In data pages V1 the size of levels is not specified in the metadata,
+    // so reset_v1 receives a view of the full page and returns the number of bytes consumed.
+    size_t reset_v1(std::basic_string_view<uint8_t> buffer, format::Encoding::type encoding, int num_values);
+    // reset_v2 is passed a view of the levels only, because the size of levels is known in advance.
+    void reset_v2(std::basic_string_view<uint8_t> encoded_levels, int num_values);
+
+    // Read a batch of n levels (the last batch may be smaller than n).
+    template <typename T>
+    size_t read_batch(int n, T *out) {
+        n = std::min(n, _num_values - _values_read);
+        if (_bit_width == 0) {
+            std::fill(out, out + n, 0);
+            _values_read += n;
+            return n;
+        }
+        return std::visit(overloaded {
+            [this, n, out] (BitReader& r) {
+                size_t n_read = r.GetBatch(_bit_width, out, n);
+                _values_read += n_read;
+                return n_read;
+            },
+            [this, n, out] (RleDecoder& r) {
+                size_t n_read = r.GetBatch(out, n);
+                _values_read += n_read;
+                return n_read;
+            },
+        }, _decoder);
+    }
+};
+
+////////////////////////////////////////////////////////////////////////////////
+
+/* Refer to the parquet documentation for the description of various encodings:
+ * https://github.com/apache/parquet-format/blob/master/Encodings.md
```

Comment:
If the documentation is also apache-licensed, and I assume it is, it's probably a good idea to copy this file to our source code in `doc/parquet/`, just in case it disappears one day.

Reply:
Done in 12d4c91.